```
Error running mint automation
+ mkdir -p 15674-dfadaf4/gopath/src/github.com/minio
+ git -C 15674-dfadaf4/gopath/src/github.com/minio clone --quiet https://github.com/minio/minio.git
+ git -C 15674-dfadaf4/gopath/src/github.com/minio/minio remote add minio https://github.com/minio/minio.git
+ git -C 15674-dfadaf4/gopath/src/github.com/minio/minio fetch --quiet minio pull/15674/head:pr15674
+ git -C 15674-dfadaf4/gopath/src/github.com/minio/minio checkout --quiet pr15674
+ GO111MODULE=on
+ GOPATH=/home/mint-auto/mint-auto/15674-dfadaf4/gopath
+ make -C 15674-dfadaf4/gopath/src/github.com/minio/minio --quiet
go: downloading github.com/minio/madmin-go v1.4.26
go: downloading github.com/minio/cli v1.23.0
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/minio/minio-go/v7 v7.0.35
go: downloading github.com/secure-io/sio-go v0.3.1
go: downloading github.com/minio/sha256-simd v1.0.0
go: downloading github.com/klauspost/compress v1.15.9
go: downloading github.com/golang-jwt/jwt/v4 v4.4.1
go: downloading github.com/buger/jsonparser v1.1.1
go: downloading github.com/fatih/color v1.13.0
go: downloading github.com/Shopify/sarama v1.35.0
go: downloading github.com/minio/sio v0.3.0
go: downloading github.com/xdg/scram v0.0.0-20180814205039-7eeb5667e42c
go: downloading golang.org/x/sys v0.0.0-20220722155257-8c9f86f7a55f
go: downloading github.com/minio/kes v0.20.0
go: downloading github.com/dustin/go-humanize v1.0.0
go: downloading github.com/minio/pkg v1.3.2
go: downloading golang.org/x/crypto v0.0.0-20220722155217-630584e8d5aa
go: downloading github.com/minio/highwayhash v1.0.2
go: downloading github.com/rs/dnscache v0.0.0-20211102005908-e0241e321417
go: downloading github.com/Azure/azure-storage-blob-go v0.10.0
go: downloading github.com/bcicen/jstream v1.0.1
go: downloading github.com/djherbis/atime v1.0.0
go: downloading github.com/cheggaaa/pb v1.0.29
go: downloading github.com/cosnicolaou/pbzip2 v1.0.1
go: downloading github.com/bits-and-blooms/bloom/v3 v3.0.1
go: downloading github.com/inconshreveable/mousetrap v1.0.0
go: downloading github.com/go-openapi/loads v0.21.1
go: downloading github.com/dchest/siphash v1.2.1
go: downloading github.com/coreos/go-oidc v2.1.0+incompatible
go: downloading github.com/klauspost/pgzip v1.2.5
go: downloading github.com/gorilla/mux v1.8.0
go: downloading cloud.google.com/go/storage v1.10.0
go: downloading github.com/felixge/fgprof v0.9.2
go: downloading github.com/cespare/xxhash/v2 v2.1.2
go: downloading github.com/google/uuid v1.3.0
go: downloading github.com/hashicorp/golang-lru v0.5.4
go: downloading github.com/miekg/dns v1.1.49
go: downloading github.com/minio/console v0.20.1
go: downloading github.com/klauspost/reedsolomon v1.10.0
go: downloading github.com/minio/dperf v0.4.2
go: downloading github.com/klauspost/readahead v1.4.0
go: downloading github.com/minio/selfupdate v0.5.0
go: downloading github.com/minio/xxml v0.0.3
go: downloading github.com/minio/zipindex v0.2.1
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/philhofer/fwd v1.1.2-0.20210722190033-5c56ac6d0bb9
go: downloading github.com/pierrec/lz4 v2.6.1+incompatible
go: downloading github.com/prometheus/client_golang v1.12.2
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading github.com/rs/cors v1.7.0
go: downloading github.com/prometheus/procfs v0.7.3
go: downloading github.com/shirou/gopsutil/v3 v3.22.7
go: downloading github.com/tinylib/msgp v1.1.7-0.20211026165309-e818a1881b0e
go: downloading github.com/prometheus/common v0.34.0
go: downloading github.com/valyala/bytebufferpool v1.0.0
go: downloading github.com/yargevad/filepathx v1.0.0
go: downloading github.com/zeebo/xxh3 v1.0.2
go: downloading go.etcd.io/etcd/api/v3 v3.5.4
go: downloading go.etcd.io/etcd/client/v3 v3.5.4
go: downloading go.uber.org/atomic v1.9.0
go: downloading golang.org/x/oauth2 v0.0.0-20220411215720-9780585627b5
go: downloading google.golang.org/api v0.81.0
go: downloading github.com/beevik/ntp v0.3.0
go: downloading github.com/go-ldap/ldap/v3 v3.4.3
go: downloading golang.org/x/time v0.0.0-20220224211638-0e9765cccd65
go: downloading go.uber.org/zap v1.21.0
go: downloading github.com/coredns/coredns v1.9.3
go: downloading github.com/ncw/directio v1.0.5
go: downloading github.com/eclipse/paho.mqtt.golang v1.4.1
go: downloading github.com/elastic/go-elasticsearch/v7 v7.17.1
go: downloading github.com/go-sql-driver/mysql v1.6.0
go: downloading github.com/gomodule/redigo v1.8.8
go: downloading github.com/lib/pq v1.10.4
go: downloading github.com/nats-io/nats.go v1.16.0
go: downloading github.com/nats-io/stan.go v0.10.2
go: downloading github.com/nsqio/go-nsq v1.0.8
go: downloading github.com/pkg/errors v0.9.1
go: downloading github.com/streadway/amqp v1.0.0
go: downloading github.com/minio/simdjson-go v0.4.2
go: downloading github.com/pquerna/cachecontrol v0.1.0
go: downloading gopkg.in/square/go-jose.v2 v2.5.1
go: downloading github.com/mattn/go-runewidth v0.0.13
go: downloading github.com/mattn/go-colorable v0.1.12
go: downloading github.com/mattn/go-isatty v0.0.14
go: downloading golang.org/x/sync v0.0.0-20220513210516-0976fa681c29
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/golang/protobuf v1.5.2
go: downloading github.com/klauspost/cpuid/v2 v2.1.0
go: downloading github.com/xdg/stringprep v1.0.0
go: downloading github.com/minio/csvparser v1.0.0
go: downloading github.com/fraugster/parquet-go v0.10.0
go: downloading github.com/alecthomas/participle v0.2.1
go: downloading golang.org/x/net v0.0.0-20220722155237-a158d28d115b
go: downloading github.com/bits-and-blooms/bitset v1.2.0
go: downloading github.com/Azure/go-ntlmssp v0.0.0-20211209120228-48547f28849e
go: downloading github.com/go-asn1-ber/asn1-ber v1.5.4
go: downloading github.com/rivo/uniseg v0.2.0
go: downloading github.com/golang/snappy v0.0.4
go: downloading github.com/go-openapi/analysis v0.21.2
go: downloading github.com/go-openapi/spec v0.20.5
go: downloading github.com/go-openapi/swag v0.21.1
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading github.com/nats-io/nuid v1.0.1
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.2-0.20181231171920-c182affec369
go: downloading go.etcd.io/etcd/client/pkg/v3 v3.5.4
go: downloading google.golang.org/grpc v1.46.2
go: downloading golang.org/x/text v0.3.7
go: downloading google.golang.org/genproto v0.0.0-20220519153652-3a47de7e79bd
go: downloading github.com/coreos/go-semver v0.3.0
go: downloading github.com/gorilla/websocket v1.5.0
go: downloading github.com/montanaflynn/stats v0.6.6
go: downloading github.com/lestrrat-go/jwx v1.2.19
go: downloading github.com/fatih/structs v1.1.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/rjeczalik/notify v0.9.2
go: downloading github.com/Azure/azure-pipeline-go v0.2.2
go: downloading github.com/mailru/easyjson v0.7.7
go: downloading github.com/go-openapi/errors v0.20.2
go: downloading github.com/go-openapi/runtime v0.23.3
go: downloading github.com/go-openapi/strfmt v0.21.2
go: downloading github.com/go-openapi/validate v0.21.0
go: downloading github.com/jessevdk/go-flags v1.5.0
go: downloading github.com/minio/mc v0.0.0-20220818165341-8c239d16aa37
go: downloading github.com/rs/xid v1.4.0
go: downloading github.com/unrolled/secure v1.10.0
go: downloading github.com/tidwall/gjson v1.14.2
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.28.0
go: downloading go.uber.org/multierr v1.8.0
go: downloading github.com/google/pprof v0.0.0-20211214055906-6f57359322fd
go: downloading github.com/tklauser/go-sysconf v0.3.10
go: downloading github.com/mattn/go-ieproxy v0.0.1
go: downloading github.com/coreos/go-systemd/v22 v22.3.2
go: downloading gopkg.in/ini.v1 v1.66.6
go: downloading github.com/minio/md5-simd v1.1.2
go: downloading github.com/oklog/ulid v1.3.1
go: downloading go.mongodb.org/mongo-driver v1.9.1
go: downloading github.com/mitchellh/mapstructure v1.4.3
go: downloading github.com/asaskevich/govalidator v0.0.0-20210307081110-f21760c49a8d
go: downloading github.com/josharian/intern v1.0.0
go: downloading github.com/tidwall/pretty v1.2.0
go: downloading github.com/tidwall/match v1.1.1
go: downloading github.com/nats-io/nkeys v0.3.0
go: downloading github.com/tklauser/numcpus v0.4.0
go: downloading cloud.google.com/go v0.100.2
go: downloading github.com/apache/thrift v0.15.0
go: downloading github.com/rcrowley/go-metrics v0.0.0-20201227073835-cf1acfcdf475
go: downloading github.com/eapache/queue v1.1.0
go: downloading github.com/eapache/go-resiliency v1.3.0
go: downloading github.com/jcmturner/gofork v1.0.0
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/hashicorp/go-multierror v1.1.1
go: downloading github.com/pierrec/lz4/v4 v4.1.15
go: downloading github.com/jcmturner/gokrb5/v8 v8.4.2
go: downloading github.com/eapache/go-xerial-snappy v0.0.0-20180814174437-776d5712da21
go: downloading cloud.google.com/go/compute v1.6.1
go: downloading github.com/docker/go-units v0.4.0
go: downloading github.com/go-openapi/jsonpointer v0.19.5
go: downloading github.com/hashicorp/errwrap v1.1.0
go: downloading github.com/go-openapi/jsonreference v0.20.0
go: downloading github.com/lestrrat-go/option v1.0.0
go: downloading github.com/lestrrat-go/iter v1.0.1
go: downloading github.com/lestrrat-go/backoff/v2 v2.0.8
go: downloading github.com/lestrrat-go/httpcc v1.0.0
go: downloading github.com/lestrrat-go/blackmagic v1.0.0
go: downloading github.com/jcmturner/dnsutils/v2 v2.0.0
go: downloading github.com/hashicorp/go-uuid v1.0.2
go: downloading github.com/jcmturner/aescts/v2 v2.0.0
go: downloading github.com/jcmturner/rpc/v2 v2.0.3
go: downloading github.com/charmbracelet/bubbletea v0.20.0
go: downloading github.com/charmbracelet/lipgloss v0.5.0
go: downloading github.com/pkg/xattr v0.4.5
go: downloading github.com/olekukonko/tablewriter v0.0.5
go: downloading github.com/posener/complete v1.2.3
go: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211
go: downloading github.com/charmbracelet/bubbles v0.10.3
go: downloading github.com/rivo/tview v0.0.0-20220216162559-96063d6082f3
go: downloading github.com/navidys/tvxwidgets v0.1.0
go: downloading gopkg.in/h2non/filetype.v1 v1.0.5
go: downloading github.com/prometheus/prom2json v1.3.1
go: downloading github.com/minio/filepath v1.0.0
go: downloading github.com/minio/colorjson v1.0.2
go: downloading github.com/google/shlex v0.0.0-20191202100458-e7afc7fbc510
go: downloading github.com/gdamore/tcell/v2 v2.4.1-0.20210905002822-f057f0a857a1
go: downloading github.com/go-stack/stack v1.8.1
go: downloading github.com/muesli/termenv v0.11.1-0.20220212125758-44cd13922739
go: downloading github.com/muesli/reflow v0.3.0
go: downloading github.com/lucasb-eyer/go-colorful v1.2.0
go: downloading github.com/containerd/console v1.0.3
go: downloading github.com/muesli/ansi v0.0.0-20211031195517-c9f0611b6c70
go: downloading github.com/gdamore/encoding v1.0.0
../../../../pkg/mod/golang.org/x/oauth2@v0.0.0-20220411215720-9780585627b5/google/default.go:17:2: cloud.google.com/go@v0.100.2: stream error: stream ID 293; INTERNAL_ERROR; received from peer
cmd/warm-backend-gcs.go:25:2: cloud.google.com/go@v0.100.2: stream error: stream ID 293; INTERNAL_ERROR; received from peer
make: *** [build] Error 1

```

--- a/cmd/bucket-replication.go
+++ b/cmd/bucket-replication.go
@@ -946,7 +946,12 @@ func replicateObject(ctx context.Context, ri ReplicateObjectInfo, objectAPI Obje
 		wg.Add(1)
 		go func(index int, tgt *TargetClient) {
 			defer wg.Done()
-			rinfos.Targets[index] = replicateObjectToTarget(ctx, ri, objectAPI, tgt)
+			if ri.OpType == replication.ObjectReplicationType {
+				// all incoming calls go through optimized path.
+				rinfos.Targets[index] = ri.replicateObject(ctx, objectAPI, tgt)
+			} else {
+				rinfos.Targets[index] = ri.replicateAll(ctx, objectAPI, tgt)
+			}
 		}(i, tgt)
 	}
 	wg.Wait()
@@ -1013,30 +1018,16 @@ func replicateObject(ctx context.Context, ri ReplicateObjectInfo, objectAPI Obje
 	}
 }
 
-// replicateObjectToTarget replicates the specified version of the object to destination bucket
+// replicateObject replicates object data for specified version of the object to destination bucket
 // The source object is then updated to reflect the replication status.
-func replicateObjectToTarget(ctx context.Context, ri ReplicateObjectInfo, objectAPI ObjectLayer, tgt *TargetClient) (rinfo replicatedTargetInfo) {
+func (ri ReplicateObjectInfo) replicateObject(ctx context.Context, objectAPI ObjectLayer, tgt *TargetClient) (rinfo replicatedTargetInfo) {
 	startTime := time.Now()
 	objInfo := ri.ObjectInfo.Clone()
 	bucket := objInfo.Bucket
 	object := objInfo.Name
-	var (
-		closeOnDefer bool
-		gr           *GetObjectReader
-		size         int64
-		err          error
-	)
 	sz, _ := objInfo.GetActualSize()
-	// set defaults for replication action based on operation being performed - actual
-	// replication action can only be determined after stat on remote. This default is
-	// needed for updating replication metrics correctly when target is offline.
-	var rAction replicationAction
-	switch ri.OpType {
-	case replication.MetadataReplicationType:
-		rAction = replicateMetadata
-	default:
-		rAction = replicateAll
-	}
+
+	rAction := replicateAll
 	rinfo = replicatedTargetInfo{
 		Size:                  sz,
 		Arn:                   tgt.ARN,
@@ -1051,6 +1042,7 @@ func replicateObjectToTarget(ctx context.Context, ri ReplicateObjectInfo, object
 		rinfo.ReplicationResynced = true
 		return
 	}
+
 	if globalBucketTargetSys.isOffline(tgt.EndpointURL()) {
 		logger.LogIf(ctx, fmt.Errorf("remote target is offline for bucket:%s arn:%s", bucket, tgt.ARN))
 		sendEvent(eventArgs{
@@ -1065,7 +1057,7 @@ func replicateObjectToTarget(ctx context.Context, ri ReplicateObjectInfo, object
 	versioned := globalBucketVersioningSys.PrefixEnabled(bucket, object)
 	versionSuspended := globalBucketVersioningSys.PrefixSuspended(bucket, object)
 
-	gr, err = objectAPI.GetObjectNInfo(ctx, bucket, object, nil, http.Header{}, readLock, ObjectOptions{
+	gr, err := objectAPI.GetObjectNInfo(ctx, bucket, object, nil, http.Header{}, readLock, ObjectOptions{
 		VersionID:        objInfo.VersionID,
 		Versioned:        versioned,
 		VersionSuspended: versionSuspended,
@@ -1082,15 +1074,159 @@ func replicateObjectToTarget(ctx context.Context, ri ReplicateObjectInfo, object
 		}
 		return
 	}
-	defer func() {
-		if closeOnDefer {
-			gr.Close()
-		}
-	}()
-	closeOnDefer = true
+	defer gr.Close()
 
 	objInfo = gr.ObjInfo
-	size, err = objInfo.GetActualSize()
+
+	size, err := objInfo.GetActualSize()
+	if err != nil {
+		logger.LogIf(ctx, err)
+		sendEvent(eventArgs{
+			EventName:  event.ObjectReplicationNotTracked,
+			BucketName: bucket,
+			Object:     objInfo,
+			Host:       "Internal: [Replication]",
+		})
+		return
+	}
+
+	if tgt.Bucket == "" {
+		logger.LogIf(ctx, fmt.Errorf("Unable to replicate object %s(%s), bucket is empty", objInfo.Name, objInfo.VersionID))
+		sendEvent(eventArgs{
+			EventName:  event.ObjectReplicationNotTracked,
+			BucketName: bucket,
+			Object:     objInfo,
+			Host:       "Internal: [Replication]",
+		})
+		return rinfo
+	}
+	defer func() {
+		if rinfo.ReplicationStatus == replication.Completed && ri.OpType == replication.ExistingObjectReplicationType && tgt.ResetID != "" {
+			rinfo.ResyncTimestamp = fmt.Sprintf("%s;%s", UTCNow().Format(http.TimeFormat), tgt.ResetID)
+			rinfo.ReplicationResynced = true
+		}
+		rinfo.Duration = time.Since(startTime)
+	}()
+
+	rinfo.ReplicationStatus = replication.Completed
+	rinfo.Size = size
+	rinfo.ReplicationAction = rAction
+	// use core client to avoid doing multipart on PUT
+	c := &miniogo.Core{Client: tgt.Client}
+
+	putOpts, err := putReplicationOpts(ctx, tgt.StorageClass, objInfo)
+	if err != nil {
+		logger.LogIf(ctx, fmt.Errorf("failed to get target for replication bucket:%s err:%w", bucket, err))
+		sendEvent(eventArgs{
+			EventName:  event.ObjectReplicationNotTracked,
+			BucketName: bucket,
+			Object:     objInfo,
+			Host:       "Internal: [Replication]",
+		})
+		return
+	}
+
+	var headerSize int
+	for k, v := range putOpts.Header() {
+		headerSize += len(k) + len(v)
+	}
+
+	opts := &bandwidth.MonitorReaderOptions{
+		Bucket:     objInfo.Bucket,
+		HeaderSize: headerSize,
+	}
+	newCtx := ctx
+	if globalBucketMonitor.IsThrottled(bucket) {
+		var cancel context.CancelFunc
+		newCtx, cancel = context.WithTimeout(ctx, throttleDeadline)
+		defer cancel()
+	}
+	r := bandwidth.NewMonitoredReader(newCtx, globalBucketMonitor, gr, opts)
+	if objInfo.isMultipart() {
+		if err := replicateObjectWithMultipart(ctx, c, tgt.Bucket, object,
+			r, objInfo, putOpts); err != nil {
+			if minio.ToErrorResponse(err).Code != "PreConditionFailed" {
+				rinfo.ReplicationStatus = replication.Failed
+				logger.LogIf(ctx, fmt.Errorf("Unable to replicate for object %s/%s(%s): %s", bucket, objInfo.Name, objInfo.VersionID, err))
+			}
+		}
+	} else {
+		if _, err = c.PutObject(ctx, tgt.Bucket, object, r, size, "", "", putOpts); err != nil {
+			if minio.ToErrorResponse(err).Code != "PreConditionFailed" {
+				rinfo.ReplicationStatus = replication.Failed
+				logger.LogIf(ctx, fmt.Errorf("Unable to replicate for object %s/%s(%s): %s", bucket, objInfo.Name, objInfo.VersionID, err))
+			}
+		}
+	}
+	return
+}
+
+// replicateAll replicates metadata for specified version of the object to destination bucket
+// if the destination version is missing it automatically does fully copy as well.
+// The source object is then updated to reflect the replication status.
+func (ri ReplicateObjectInfo) replicateAll(ctx context.Context, objectAPI ObjectLayer, tgt *TargetClient) (rinfo replicatedTargetInfo) {
+	startTime := time.Now()
+	objInfo := ri.ObjectInfo.Clone()
+	bucket := objInfo.Bucket
+	object := objInfo.Name
+	sz, _ := objInfo.GetActualSize()
+
+	// set defaults for replication action based on operation being performed - actual
+	// replication action can only be determined after stat on remote. This default is
+	// needed for updating replication metrics correctly when target is offline.
+	rAction := replicateMetadata
+
+	rinfo = replicatedTargetInfo{
+		Size:                  sz,
+		Arn:                   tgt.ARN,
+		PrevReplicationStatus: objInfo.TargetReplicationStatus(tgt.ARN),
+		ReplicationStatus:     replication.Failed,
+		OpType:                ri.OpType,
+		ReplicationAction:     rAction,
+	}
+
+	if ri.ObjectInfo.TargetReplicationStatus(tgt.ARN) == replication.Completed && !ri.ExistingObjResync.Empty() && !ri.ExistingObjResync.mustResyncTarget(tgt.ARN) {
+		rinfo.ReplicationStatus = replication.Completed
+		rinfo.ReplicationResynced = true
+		return
+	}
+
+	if globalBucketTargetSys.isOffline(tgt.EndpointURL()) {
+		logger.LogIf(ctx, fmt.Errorf("remote target is offline for bucket:%s arn:%s", bucket, tgt.ARN))
+		sendEvent(eventArgs{
+			EventName:  event.ObjectReplicationNotTracked,
+			BucketName: bucket,
+			Object:     objInfo,
+			Host:       "Internal: [Replication]",
+		})
+		return
+	}
+
+	versioned := globalBucketVersioningSys.PrefixEnabled(bucket, object)
+	versionSuspended := globalBucketVersioningSys.PrefixSuspended(bucket, object)
+
+	gr, err := objectAPI.GetObjectNInfo(ctx, bucket, object, nil, http.Header{}, readLock, ObjectOptions{
+		VersionID:        objInfo.VersionID,
+		Versioned:        versioned,
+		VersionSuspended: versionSuspended,
+	})
+	if err != nil {
+		if !isErrObjectNotFound(err) {
+			sendEvent(eventArgs{
+				EventName:  event.ObjectReplicationNotTracked,
+				BucketName: bucket,
+				Object:     objInfo,
+				Host:       "Internal: [Replication]",
+			})
+			logger.LogIf(ctx, fmt.Errorf("Unable to update replicate metadata for %s/%s(%s): %w", bucket, object, objInfo.VersionID, err))
+		}
+		return
+	}
+	defer gr.Close()
+
+	objInfo = gr.ObjInfo
+
+	size, err := objInfo.GetActualSize()
 	if err != nil {
 		logger.LogIf(ctx, err)
 		sendEvent(eventArgs{
@@ -1149,8 +1285,6 @@ func replicateObjectToTarget(ctx context.Context, ri ReplicateObjectInfo, object
 				// as Completed.
 				//
 				// Note: Replication Stats would have been updated despite metadata update failure.
-				gr.Close()
-				closeOnDefer = false
 				rinfo.ReplicationAction = rAction
 				rinfo.ReplicationStatus = replication.Completed
 			}
@@ -1221,8 +1355,6 @@ func replicateObjectToTarget(ctx context.Context, ri ReplicateObjectInfo, object
 			}
 		}
 	}
-	gr.Close()
-	closeOnDefer = false
 	return
 }
 

--- a/cmd/erasure-multipart.go
+++ b/cmd/erasure-multipart.go
@@ -327,6 +327,24 @@ func (er erasureObjects) ListMultipartUploads(ctx context.Context, bucket, objec
 // disks. `uploads.json` carries metadata regarding on-going multipart
 // operation(s) on the object.
 func (er erasureObjects) newMultipartUpload(ctx context.Context, bucket string, object string, opts ObjectOptions) (*NewMultipartUploadResult, error) {
+	if opts.CheckPrecondFn != nil {
+		// Lock the object before reading.
+		lk := er.NewNSLock(bucket, object)
+		lkctx, err := lk.GetRLock(ctx, globalOperationTimeout)
+		if err != nil {
+			return nil, err
+		}
+		rctx := lkctx.Context()
+		obj, err := er.getObjectInfo(rctx, bucket, object, opts)
+		lk.RUnlock(lkctx.Cancel)
+		if err != nil {
+			return nil, err
+		}
+		if opts.CheckPrecondFn(obj) {
+			return nil, PreConditionFailed{}
+		}
+	}
+
 	userDefined := cloneMSS(opts.UserDefined)
 
 	onlineDisks := er.getDisks()

--- a/cmd/erasure-object.go
+++ b/cmd/erasure-object.go
@@ -915,6 +915,16 @@ func (er erasureObjects) PutObject(ctx context.Context, bucket string, object st
 func (er erasureObjects) putObject(ctx context.Context, bucket string, object string, r *PutObjReader, opts ObjectOptions) (objInfo ObjectInfo, err error) {
 	auditObjectErasureSet(ctx, object, &er)
 
+	if opts.CheckPrecondFn != nil {
+		obj, err := er.getObjectInfo(ctx, bucket, object, opts)
+		if err != nil {
+			return objInfo, err
+		}
+		if opts.CheckPrecondFn(obj) {
+			return objInfo, PreConditionFailed{}
+		}
+	}
+
 	data := r.Reader
 
 	userDefined := cloneMSS(opts.UserDefined)
@@ -1146,11 +1156,9 @@ func (er erasureObjects) putObject(ctx context.Context, bucket string, object st
 		})
 	}
 
-	if userDefined["etag"] == "" {
-		userDefined["etag"] = r.MD5CurrentHexString()
-		if opts.PreserveETag != "" {
-			userDefined["etag"] = opts.PreserveETag
-		}
+	userDefined["etag"] = r.MD5CurrentHexString()
+	if opts.PreserveETag != "" {
+		userDefined["etag"] = opts.PreserveETag
 	}
 
 	// Guess content-type from the extension if possible.

--- a/cmd/object-api-options.go
+++ b/cmd/object-api-options.go
@@ -283,13 +283,11 @@ func putOpts(ctx context.Context, r *http.Request, bucket, object string, metada
 		}
 	}
 
-	etag := strings.TrimSpace(r.Header.Get(xhttp.MinIOSourceETag))
-	if etag != "" {
-		if metadata == nil {
-			metadata = make(map[string]string, 1)
-		}
-		metadata["etag"] = etag
+	if metadata == nil {
+		metadata = make(map[string]string)
 	}
+
+	etag := strings.TrimSpace(r.Header.Get(xhttp.MinIOSourceETag))
 
 	wantCRC, err := hash.GetContentChecksum(r)
 	if err != nil {
@@ -310,6 +308,7 @@ func putOpts(ctx context.Context, r *http.Request, bucket, object string, metada
 			Versioned:            versioned,
 			VersionSuspended:     versionSuspended,
 			MTime:                mtime,
+			PreserveETag:         etag,
 			WantChecksum:         wantCRC,
 		}, nil
 	}

--- a/cmd/object-handlers-common.go
+++ b/cmd/object-handlers-common.go
@@ -130,6 +130,54 @@ func checkCopyObjectPreconditions(ctx context.Context, w http.ResponseWriter, r 
 	return false
 }
 
+// Validates the preconditions. Returns true if PUT operation should not proceed.
+// Preconditions supported are:
+//
+//	x-minio-source-mtime
+//	x-minio-source-etag
+func checkPreconditionsPUT(ctx context.Context, w http.ResponseWriter, r *http.Request, objInfo ObjectInfo, opts ObjectOptions) bool {
+	// Return false for methods other than PUT.
+	if r.Method != http.MethodPut {
+		return false
+	}
+	// If the object doesn't have a modtime (IsZero), or the modtime
+	// is obviously garbage (Unix time == 0), then ignore modtimes
+	// and don't process the If-Modified-Since header.
+	if objInfo.ModTime.IsZero() || objInfo.ModTime.Equal(time.Unix(0, 0)) {
+		return false
+	}
+
+	// If top level is a delete marker proceed to upload.
+	if objInfo.DeleteMarker {
+		return false
+	}
+
+	// Headers to be set of object content is not going to be written to the client.
+	writeHeaders := func() {
+		// set common headers
+		setCommonHeaders(w)
+
+		// set object-related metadata headers
+		w.Header().Set(xhttp.LastModified, objInfo.ModTime.UTC().Format(http.TimeFormat))
+
+		if objInfo.ETag != "" {
+			w.Header()[xhttp.ETag] = []string{"\"" + objInfo.ETag + "\""}
+		}
+	}
+
+	etagMatch := opts.PreserveETag != "" && isETagEqual(objInfo.ETag, opts.PreserveETag)
+	vidMatch := opts.VersionID != "" && opts.VersionID == objInfo.VersionID
+	mtimeMatch := !opts.MTime.IsZero() && objInfo.ModTime.Unix() >= opts.MTime.Unix()
+	if etagMatch && vidMatch && mtimeMatch {
+		writeHeaders()
+		writeErrorResponse(ctx, w, errorCodes.ToAPIErr(ErrPreconditionFailed), r.URL)
+		return true
+	}
+
+	// Object content should be persisted.
+	return false
+}
+
 // Validates the preconditions. Returns true if GET/HEAD operation should not proceed.
 // Preconditions supported are:
 //

--- a/cmd/object-handlers.go
+++ b/cmd/object-handlers.go
@@ -1786,6 +1786,18 @@ func (api objectAPIHandlers) PutObjectHandler(w http.ResponseWriter, r *http.Req
 	}
 	opts.IndexCB = idxCb
 
+	if !opts.MTime.IsZero() && opts.PreserveETag != "" {
+		opts.CheckPrecondFn = func(oi ObjectInfo) bool {
+			if objectAPI.IsEncryptionSupported() {
+				if _, err := DecryptObjectInfo(&oi, r); err != nil {
+					writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL)
+					return true
+				}
+			}
+			return checkPreconditionsPUT(ctx, w, r, oi, opts)
+		}
+	}
+
 	if api.CacheAPI() != nil {
 		putObject = api.CacheAPI().PutObject
 	}

--- a/cmd/object-multipart-handlers.go
+++ b/cmd/object-multipart-handlers.go
@@ -176,6 +176,18 @@ func (api objectAPIHandlers) NewMultipartUploadHandler(w http.ResponseWriter, r 
 		return
 	}
 
+	if !opts.MTime.IsZero() && opts.PreserveETag != "" {
+		opts.CheckPrecondFn = func(oi ObjectInfo) bool {
+			if objectAPI.IsEncryptionSupported() {
+				if _, err := DecryptObjectInfo(&oi, r); err != nil {
+					writeErrorResponse(ctx, w, toAPIError(ctx, err), r.URL)
+					return true
+				}
+			}
+			return checkPreconditionsPUT(ctx, w, r, oi, opts)
+		}
+	}
+
 	checksumType := hash.NewChecksumType(r.Header.Get(xhttp.AmzChecksumAlgo))
 	if checksumType.Is(hash.ChecksumInvalid) {
 		writeErrorResponse(ctx, w, errorCodes.ToAPIErr(ErrInvalidRequestParameter), r.URL)

